File: packages/firebase_auth/firebase_auth/android/src/main/java/io/flutter/plugins/firebase/auth/PigeonParser.java
/*
 * Copyright 2023, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */

package io.flutter.plugins.firebase.auth;

import android.net.Uri;
import androidx.annotation.NonNull;
import com.google.firebase.auth.ActionCodeEmailInfo;
import com.google.firebase.auth.ActionCodeInfo;
import com.google.firebase.auth.ActionCodeResult;
import com.google.firebase.auth.ActionCodeSettings;
import com.google.firebase.auth.AdditionalUserInfo;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.EmailAuthProvider;
import com.google.firebase.auth.FacebookAuthProvider;
import com.google.firebase.auth.FirebaseAuthProvider;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.FirebaseUserMetadata;
import com.google.firebase.auth.GetTokenResult;
import com.google.firebase.auth.GithubAuthProvider;
import com.google.firebase.auth.GoogleAuthProvider;
import com.google.firebase.auth.MultiFactorInfo;
import com.google.firebase.auth.OAuthCredential;
import com.google.firebase.auth.OAuthProvider;
import com.google.firebase.auth.PhoneAuthProvider;
import com.google.firebase.auth.PhoneMultiFactorInfo;
import com.google.firebase.auth.PlayGamesAuthProvider;
import com.google.firebase.auth.TwitterAuthProvider;
import com.google.firebase.auth.UserInfo;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class PigeonParser {
  static List<Object> manuallyToList(
      GeneratedAndroidFirebaseAuth.PigeonUserDetails pigeonUserDetails) {
    List<Object> output = new ArrayList<>();
    output.add(pigeonUserDetails.getUserInfo().toList());
    output.add(pigeonUserDetails.getProviderData());
    return output;
  }

  static GeneratedAndroidFirebaseAuth.PigeonUserCredential parseAuthResult(
      @NonNull AuthResult authResult) {
    GeneratedAndroidFirebaseAuth.PigeonUserCredential.Builder builder =
        new GeneratedAndroidFirebaseAuth.PigeonUserCredential.Builder();

    builder.setAdditionalUserInfo(parseAdditionalUserInfo(authResult.getAdditionalUserInfo()));
    builder.setCredential(parseAuthCredential(authResult.getCredential()));
    builder.setUser(parseFirebaseUser(authResult.getUser()));

    return builder.build();
  }

  private static GeneratedAndroidFirebaseAuth.PigeonAdditionalUserInfo parseAdditionalUserInfo(
      AdditionalUserInfo additionalUserInfo) {
    if (additionalUserInfo == null) {
      return null;
    }

    GeneratedAndroidFirebaseAuth.PigeonAdditionalUserInfo.Builder builder =
        new GeneratedAndroidFirebaseAuth.PigeonAdditionalUserInfo.Builder();

    builder.setIsNewUser(additionalUserInfo.isNewUser());
    builder.setProfile(additionalUserInfo.getProfile());
    builder.setProviderId(additionalUserInfo.getProviderId());
    builder.setUsername(additionalUserInfo.getUsername());

    return builder.build();
  }

  static GeneratedAndroidFirebaseAuth.PigeonAuthCredential parseAuthCredential(
      AuthCredential authCredential) {
    if (authCredential == null) {
      return null;
    }

    int authCredentialHashCode = authCredential.hashCode();
    FlutterFirebaseAuthPlugin.authCredentials.put(authCredentialHashCode, authCredential);

    GeneratedAndroidFirebaseAuth.PigeonAuthCredential.Builder builder =
        new GeneratedAndroidFirebaseAuth.PigeonAuthCredential.Builder();

    builder.setProviderId(authCredential.getProvider());
    builder.setSignInMethod(authCredential.getSignInMethod());
    builder.setNativeId((long) authCredentialHashCode);
    if (authCredential instanceof OAuthCredential) {
      builder.setAccessToken(((OAuthCredential) authCredential).getAccessToken());
    }

    return builder.build();
  }

  static GeneratedAndroidFirebaseAuth.PigeonUserDetails parseFirebaseUser(
      FirebaseUser firebaseUser) {
    if (firebaseUser == null) {
      return null;
    }

    GeneratedAndroidFirebaseAuth.PigeonUserDetails.Builder builder =
        new GeneratedAndroidFirebaseAuth.PigeonUserDetails.Builder();

    GeneratedAndroidFirebaseAuth.PigeonUserInfo.Builder builderInfo =
        new GeneratedAndroidFirebaseAuth.PigeonUserInfo.Builder();

    builderInfo.setDisplayName(firebaseUser.getDisplayName());
    builderInfo.setEmail(firebaseUser.getEmail());
    builderInfo.setIsEmailVerified(firebaseUser.isEmailVerified());
    builderInfo.setIsAnonymous(firebaseUser.isAnonymous());

    final FirebaseUserMetadata userMetadata = firebaseUser.getMetadata();
    if (userMetadata != null) {
      builderInfo.setCreationTimestamp(firebaseUser.getMetadata().getCreationTimestamp());
      builderInfo.setLastSignInTimestamp(firebaseUser.getMetadata().getLastSignInTimestamp());
    }
    builderInfo.setPhoneNumber(firebaseUser.getPhoneNumber());
    builderInfo.setPhotoUrl(parsePhotoUrl(firebaseUser.getPhotoUrl()));
    builderInfo.setUid(firebaseUser.getUid());
    builderInfo.setTenantId(firebaseUser.getTenantId());

    builder.setUserInfo(builderInfo.build());
    builder.setProviderData(parseUserInfoList(firebaseUser.getProviderData()));

    return builder.build();
  }

  private static List<Map<Object, Object>> parseUserInfoList(
      List<? extends UserInfo> userInfoList) {
    List<Map<Object, Object>> output = new ArrayList<>();

    if (userInfoList == null) {
      return null;
    }

    for (UserInfo userInfo : new ArrayList<UserInfo>(userInfoList)) {
      if (userInfo == null) {
        continue;
      }
      if (!FirebaseAuthProvider.PROVIDER_ID.equals(userInfo.getProviderId())) {
        output.add(parseUserInfoToMap(userInfo));
      }
    }

    return output;
  }

  private static Map<Object, Object> parseUserInfoToMap(UserInfo userInfo) {
    Map<Object, Object> output = new HashMap<>();
    output.put("displayName", userInfo.getDisplayName());
    output.put("email", userInfo.getEmail());
    output.put("isEmailVerified", userInfo.isEmailVerified());
    output.put("phoneNumber", userInfo.getPhoneNumber());
    output.put("photoUrl", parsePhotoUrl(userInfo.getPhotoUrl()));
    // Can be null on Emulator
    output.put("uid", userInfo.getUid() == null ? "" : userInfo.getUid());
    output.put("providerId", userInfo.getProviderId());
    output.put("isAnonymous", false);
    return output;
  }

  private static String parsePhotoUrl(Uri photoUri) {
    if (photoUri == null) {
      return null;
    }

    String photoUrl = photoUri.toString();

    // Return null if the URL is an empty string
    return "".equals(photoUrl) ? null : photoUrl;
  }

  @SuppressWarnings("ConstantConditions")
  static AuthCredential getCredential(Map<String, Object> credentialMap) {
    // If the credential map contains a token, it means a native one has been stored
    if (credentialMap.get(Constants.TOKEN) != null) {
      int token = (int) credentialMap.get(Constants.TOKEN);
      AuthCredential credential = FlutterFirebaseAuthPlugin.authCredentials.get(token);

      if (credential == null) {
        throw FlutterFirebaseAuthPluginException.invalidCredential();
      }

      return credential;
    }

    String signInMethod =
        (String) Objects.requireNonNull(credentialMap.get(Constants.SIGN_IN_METHOD));
    String secret = (String) credentialMap.get(Constants.SECRET);
    String idToken = (String) credentialMap.get(Constants.ID_TOKEN);
    String accessToken = (String) credentialMap.get(Constants.ACCESS_TOKEN);
    String rawNonce = (String) credentialMap.get(Constants.RAW_NONCE);

    switch (signInMethod) {
      case Constants.SIGN_IN_METHOD_PASSWORD:
        return EmailAuthProvider.getCredential(
            (String) Objects.requireNonNull(credentialMap.get(Constants.EMAIL)),
            Objects.requireNonNull(secret));
      case Constants.SIGN_IN_METHOD_EMAIL_LINK:
        return EmailAuthProvider.getCredentialWithLink(
            (String) Objects.requireNonNull(credentialMap.get(Constants.EMAIL)),
            (String) Objects.requireNonNull(credentialMap.get(Constants.EMAIL_LINK)));
      case Constants.SIGN_IN_METHOD_FACEBOOK:
        return FacebookAuthProvider.getCredential(Objects.requireNonNull(accessToken));
      case Constants.SIGN_IN_METHOD_GOOGLE:
        return GoogleAuthProvider.getCredential(idToken, accessToken);
      case Constants.SIGN_IN_METHOD_TWITTER:
        return TwitterAuthProvider.getCredential(
            Objects.requireNonNull(accessToken), Objects.requireNonNull(secret));
      case Constants.SIGN_IN_METHOD_GITHUB:
        return GithubAuthProvider.getCredential(Objects.requireNonNull(accessToken));
      case Constants.SIGN_IN_METHOD_PHONE:
        {
          String verificationId =
              (String) Objects.requireNonNull(credentialMap.get(Constants.VERIFICATION_ID));
          String smsCode = (String) Objects.requireNonNull(credentialMap.get(Constants.SMS_CODE));
          return PhoneAuthProvider.getCredential(verificationId, smsCode);
        }
      case Constants.SIGN_IN_METHOD_OAUTH:
        {
          String providerId =
              (String) Objects.requireNonNull(credentialMap.get(Constants.PROVIDER_ID));
          OAuthProvider.CredentialBuilder builder = OAuthProvider.newCredentialBuilder(providerId);
          if (accessToken != null) {
            builder.setAccessToken(accessToken);
          }
          if (rawNonce == null) {
            builder.setIdToken(Objects.requireNonNull(idToken));
          } else {
            builder.setIdTokenWithRawNonce(Objects.requireNonNull(idToken), rawNonce);
          }

          return builder.build();
        }
      case Constants.SIGN_IN_METHOD_PLAY_GAMES:
        {
          String serverAuthCode =
              (String) Objects.requireNonNull(credentialMap.get(Constants.SERVER_AUTH_CODE));
          return PlayGamesAuthProvider.getCredential(serverAuthCode);
        }
      default:
        return null;
    }
  }

  static ActionCodeSettings getActionCodeSettings(
      @NonNull GeneratedAndroidFirebaseAuth.PigeonActionCodeSettings pigeonActionCodeSettings) {
    ActionCodeSettings.Builder builder = ActionCodeSettings.newBuilder();

    builder.setUrl(pigeonActionCodeSettings.getUrl());

    if (pigeonActionCodeSettings.getDynamicLinkDomain() != null) {
      builder.setDynamicLinkDomain(pigeonActionCodeSettings.getDynamicLinkDomain());
    }

    builder.setHandleCodeInApp(pigeonActionCodeSettings.getHandleCodeInApp());

    if (pigeonActionCodeSettings.getAndroidPackageName() != null) {
      builder.setAndroidPackageName(
          pigeonActionCodeSettings.getAndroidPackageName(),
          pigeonActionCodeSettings.getAndroidInstallApp(),
          pigeonActionCodeSettings.getAndroidMinimumVersion());
    }

    if (pigeonActionCodeSettings.getIOSBundleId() != null) {
      builder.setIOSBundleId(pigeonActionCodeSettings.getIOSBundleId());
    }

    return builder.build();
  }

  static List<GeneratedAndroidFirebaseAuth.PigeonMultiFactorInfo> multiFactorInfoToPigeon(
      List<MultiFactorInfo> hints) {
    List<GeneratedAndroidFirebaseAuth.PigeonMultiFactorInfo> pigeonHints = new ArrayList<>();
    for (MultiFactorInfo info : hints) {
      if (info instanceof PhoneMultiFactorInfo) {
        pigeonHints.add(
            new GeneratedAndroidFirebaseAuth.PigeonMultiFactorInfo.Builder()
                .setPhoneNumber(((PhoneMultiFactorInfo) info).getPhoneNumber())
                .setDisplayName(info.getDisplayName())
                .setEnrollmentTimestamp((double) info.getEnrollmentTimestamp())
                .setUid(info.getUid())
                .setFactorId(info.getFactorId())
                .build());

      } else {
        pigeonHints.add(
            new GeneratedAndroidFirebaseAuth.PigeonMultiFactorInfo.Builder()
                .setDisplayName(info.getDisplayName())
                .setEnrollmentTimestamp((double) info.getEnrollmentTimestamp())
                .setUid(info.getUid())
                .setFactorId(info.getFactorId())
                .build());
      }
    }
    return pigeonHints;
  }

  static List<List<Object>> multiFactorInfoToMap(List<MultiFactorInfo> hints) {
    List<List<Object>> pigeonHints = new ArrayList<>();
    for (GeneratedAndroidFirebaseAuth.PigeonMultiFactorInfo info : multiFactorInfoToPigeon(hints)) {
      pigeonHints.add(info.toList());
    }
    return pigeonHints;
  }

  static GeneratedAndroidFirebaseAuth.PigeonActionCodeInfo parseActionCodeResult(
      @NonNull ActionCodeResult actionCodeResult) {
    GeneratedAndroidFirebaseAuth.PigeonActionCodeInfo.Builder builder =
        new GeneratedAndroidFirebaseAuth.PigeonActionCodeInfo.Builder();
    GeneratedAndroidFirebaseAuth.PigeonActionCodeInfoData.Builder builderData =
        new GeneratedAndroidFirebaseAuth.PigeonActionCodeInfoData.Builder();

    int operation = actionCodeResult.getOperation();

    switch (operation) {
      case ActionCodeResult.PASSWORD_RESET:
        builder.setOperation(GeneratedAndroidFirebaseAuth.ActionCodeInfoOperation.PASSWORD_RESET);
        break;
      case ActionCodeResult.VERIFY_EMAIL:
        builder.setOperation(GeneratedAndroidFirebaseAuth.ActionCodeInfoOperation.VERIFY_EMAIL);
        break;
      case ActionCodeResult.RECOVER_EMAIL:
        builder.setOperation(GeneratedAndroidFirebaseAuth.ActionCodeInfoOperation.RECOVER_EMAIL);
        break;
      case ActionCodeResult.SIGN_IN_WITH_EMAIL_LINK:
        builder.setOperation(GeneratedAndroidFirebaseAuth.ActionCodeInfoOperation.EMAIL_SIGN_IN);
        break;
      case ActionCodeResult.VERIFY_BEFORE_CHANGE_EMAIL:
        builder.setOperation(
            GeneratedAndroidFirebaseAuth.ActionCodeInfoOperation.VERIFY_AND_CHANGE_EMAIL);
        break;
      case ActionCodeResult.REVERT_SECOND_FACTOR_ADDITION:
        builder.setOperation(
            GeneratedAndroidFirebaseAuth.ActionCodeInfoOperation.REVERT_SECOND_FACTOR_ADDITION);
        break;
    }

    ActionCodeInfo actionCodeInfo = actionCodeResult.getInfo();

    if (actionCodeInfo != null && operation == ActionCodeResult.VERIFY_EMAIL
        || operation == ActionCodeResult.PASSWORD_RESET) {
      builderData.setEmail(actionCodeInfo.getEmail());
    } else if (operation == ActionCodeResult.RECOVER_EMAIL
        || operation == ActionCodeResult.VERIFY_BEFORE_CHANGE_EMAIL) {
      ActionCodeEmailInfo actionCodeEmailInfo =
          (ActionCodeEmailInfo) Objects.requireNonNull(actionCodeInfo);
      builderData.setEmail(actionCodeEmailInfo.getEmail());
      builderData.setPreviousEmail(actionCodeEmailInfo.getPreviousEmail());
    }

    builder.setData(builderData.build());

    return builder.build();
  }

  static GeneratedAndroidFirebaseAuth.PigeonIdTokenResult parseTokenResult(
      @NonNull GetTokenResult tokenResult) {
    final GeneratedAndroidFirebaseAuth.PigeonIdTokenResult.Builder builder =
        new GeneratedAndroidFirebaseAuth.PigeonIdTokenResult.Builder();

    builder.setToken(tokenResult.getToken());
    builder.setSignInProvider(tokenResult.getSignInProvider());
    builder.setAuthTimestamp(tokenResult.getAuthTimestamp() * 1000);
    builder.setExpirationTimestamp(tokenResult.getExpirationTimestamp() * 1000);
    builder.setIssuedAtTimestamp(tokenResult.getIssuedAtTimestamp() * 1000);
    builder.setClaims(tokenResult.getClaims());
    builder.setSignInSecondFactor(tokenResult.getSignInSecondFactor());

    return builder.build();
  }
}

File: packages/firebase_auth/firebase_auth/android/src/main/kotlin/io/flutter/plugins/firebase/auth/PigeonParser.kt
/*
 * Copyright 2023, the Chromium project authors.  Please see the AUTHORS file
 * for details. All rights reserved. Use of this source code is governed by a
 * BSD-style license that can be found in the LICENSE file.
 */
package io.flutter.plugins.firebase.auth

import android.net.Uri
import com.google.firebase.auth.ActionCodeEmailInfo
import com.google.firebase.auth.ActionCodeResult
import com.google.firebase.auth.ActionCodeSettings
import com.google.firebase.auth.AdditionalUserInfo
import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.EmailAuthProvider
import com.google.firebase.auth.FacebookAuthProvider
import com.google.firebase.auth.FirebaseAuthProvider
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GetTokenResult
import com.google.firebase.auth.GithubAuthProvider
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.MultiFactorInfo
import com.google.firebase.auth.OAuthCredential
import com.google.firebase.auth.OAuthProvider
import com.google.firebase.auth.PhoneAuthProvider
import com.google.firebase.auth.PhoneMultiFactorInfo
import com.google.firebase.auth.PlayGamesAuthProvider
import com.google.firebase.auth.TwitterAuthProvider
import com.google.firebase.auth.UserInfo

object PigeonParser {
  fun manuallyToList(pigeonUserDetails: InternalUserDetails): List<Any?> {
    return listOf(pigeonUserDetails.userInfo.toList(), pigeonUserDetails.providerData)
  }

  fun parseAuthResult(authResult: AuthResult): InternalUserCredential {
    return InternalUserCredential(
        additionalUserInfo = parseAdditionalUserInfo(authResult.additionalUserInfo),
        credential = parseAuthCredential(authResult.credential),
        user = parseFirebaseUser(authResult.user))
  }

  private fun parseAdditionalUserInfo(
      additionalUserInfo: AdditionalUserInfo?
  ): InternalAdditionalUserInfo? {
    if (additionalUserInfo == null) {
      return null
    }

    return InternalAdditionalUserInfo(
        isNewUser = additionalUserInfo.isNewUser,
        profile = additionalUserInfo.profile as Map<String?, Any?>?,
        providerId = additionalUserInfo.providerId,
        username = additionalUserInfo.username)
  }

  fun parseAuthCredential(authCredential: AuthCredential?): InternalAuthCredential? {
    if (authCredential == null) {
      return null
    }

    val authCredentialHashCode = authCredential.hashCode()
    FlutterFirebaseAuthPlugin.authCredentials[authCredentialHashCode] = authCredential

    return InternalAuthCredential(
        providerId = authCredential.provider,
        signInMethod = authCredential.signInMethod,
        nativeId = authCredentialHashCode.toLong(),
        accessToken = (authCredential as? OAuthCredential)?.accessToken)
  }

  fun parseFirebaseUser(firebaseUser: FirebaseUser?): InternalUserDetails? {
    if (firebaseUser == null) {
      return null
    }

    val userMetadata = firebaseUser.metadata
    val userInfo =
        InternalUserInfo(
            displayName = firebaseUser.displayName,
            email = firebaseUser.email,
            isEmailVerified = firebaseUser.isEmailVerified,
            isAnonymous = firebaseUser.isAnonymous,
            creationTimestamp = userMetadata?.creationTimestamp,
            lastSignInTimestamp = userMetadata?.lastSignInTimestamp,
            phoneNumber = firebaseUser.phoneNumber,
            photoUrl = parsePhotoUrl(firebaseUser.photoUrl),
            uid = firebaseUser.uid,
            tenantId = firebaseUser.tenantId)

    return InternalUserDetails(
        userInfo = userInfo, providerData = parseUserInfoList(firebaseUser.providerData))
  }

  private fun parseUserInfoList(userInfoList: List<UserInfo>?): List<Map<Any?, Any?>?> {
    val output = ArrayList<Map<Any?, Any?>?>()
    if (userInfoList == null) {
      return output
    }

    for (userInfo in ArrayList(userInfoList)) {
      if (userInfo == null) {
        continue
      }
      if (FirebaseAuthProvider.PROVIDER_ID != userInfo.providerId) {
        output.add(parseUserInfoToMap(userInfo))
      }
    }

    return output
  }

  private fun parseUserInfoToMap(userInfo: UserInfo): Map<Any?, Any?> {
    return mapOf(
        "displayName" to userInfo.displayName,
        "email" to userInfo.email,
        "isEmailVerified" to userInfo.isEmailVerified,
        "phoneNumber" to userInfo.phoneNumber,
        "photoUrl" to parsePhotoUrl(userInfo.photoUrl),
        "uid" to (userInfo.uid ?: ""),
        "providerId" to userInfo.providerId,
        "isAnonymous" to false)
  }

  private fun parsePhotoUrl(photoUri: Uri?): String? {
    if (photoUri == null) {
      return null
    }

    val photoUrl = photoUri.toString()
    return if (photoUrl == "") null else photoUrl
  }

  fun getCredential(credentialMap: Map<String?, Any?>): AuthCredential? {
    if (credentialMap[Constants.TOKEN] != null) {
      val token = (credentialMap[Constants.TOKEN] as Number).toInt()
      return FlutterFirebaseAuthPlugin.authCredentials[token]
          ?: throw FlutterFirebaseAuthPluginException.invalidCredential()
    }

    val signInMethod = credentialMap[Constants.SIGN_IN_METHOD] as String
    val secret = credentialMap[Constants.SECRET] as String?
    val idToken = credentialMap[Constants.ID_TOKEN] as String?
    val accessToken = credentialMap[Constants.ACCESS_TOKEN] as String?
    val rawNonce = credentialMap[Constants.RAW_NONCE] as String?

    return when (signInMethod) {
      Constants.SIGN_IN_METHOD_PASSWORD ->
          EmailAuthProvider.getCredential(credentialMap[Constants.EMAIL] as String, secret!!)
      Constants.SIGN_IN_METHOD_EMAIL_LINK ->
          EmailAuthProvider.getCredentialWithLink(
              credentialMap[Constants.EMAIL] as String,
              credentialMap[Constants.EMAIL_LINK] as String)
      Constants.SIGN_IN_METHOD_FACEBOOK -> FacebookAuthProvider.getCredential(accessToken!!)
      Constants.SIGN_IN_METHOD_GOOGLE -> GoogleAuthProvider.getCredential(idToken, accessToken)
      Constants.SIGN_IN_METHOD_TWITTER -> TwitterAuthProvider.getCredential(accessToken!!, secret!!)
      Constants.SIGN_IN_METHOD_GITHUB -> GithubAuthProvider.getCredential(accessToken!!)
      Constants.SIGN_IN_METHOD_PHONE -> {
        val verificationId = credentialMap[Constants.VERIFICATION_ID] as String
        val smsCode = credentialMap[Constants.SMS_CODE] as String
        PhoneAuthProvider.getCredential(verificationId, smsCode)
      }
      Constants.SIGN_IN_METHOD_OAUTH -> {
        val providerId = credentialMap[Constants.PROVIDER_ID] as String
        val builder = OAuthProvider.newCredentialBuilder(providerId)
        if (accessToken != null) {
          builder.setAccessToken(accessToken)
        }
        if (rawNonce == null) {
          builder.setIdToken(idToken!!)
        } else {
          builder.setIdTokenWithRawNonce(idToken!!, rawNonce)
        }
        builder.build()
      }
      Constants.SIGN_IN_METHOD_PLAY_GAMES -> {
        val serverAuthCode = credentialMap[Constants.SERVER_AUTH_CODE] as String
        PlayGamesAuthProvider.getCredential(serverAuthCode)
      }
      else -> null
    }
  }

  fun getActionCodeSettings(
      pigeonActionCodeSettings: InternalActionCodeSettings
  ): ActionCodeSettings {
    val builder = ActionCodeSettings.newBuilder()
    builder.setUrl(pigeonActionCodeSettings.url)

    if (pigeonActionCodeSettings.dynamicLinkDomain != null) {
      builder.setDynamicLinkDomain(pigeonActionCodeSettings.dynamicLinkDomain)
    }

    if (pigeonActionCodeSettings.linkDomain != null) {
      builder.setLinkDomain(pigeonActionCodeSettings.linkDomain)
    }

    builder.setHandleCodeInApp(pigeonActionCodeSettings.handleCodeInApp)

    if (pigeonActionCodeSettings.androidPackageName != null) {
      builder.setAndroidPackageName(
          pigeonActionCodeSettings.androidPackageName,
          pigeonActionCodeSettings.androidInstallApp,
          pigeonActionCodeSettings.androidMinimumVersion)
    }

    if (pigeonActionCodeSettings.iOSBundleId != null) {
      builder.setIOSBundleId(pigeonActionCodeSettings.iOSBundleId)
    }

    return builder.build()
  }

  fun multiFactorInfoToPigeon(hints: List<MultiFactorInfo>): List<InternalMultiFactorInfo> {
    val pigeonHints = ArrayList<InternalMultiFactorInfo>()
    for (info in hints) {
      if (info is PhoneMultiFactorInfo) {
        pigeonHints.add(
            InternalMultiFactorInfo(
                phoneNumber = info.phoneNumber,
                displayName = info.displayName,
                enrollmentTimestamp = info.enrollmentTimestamp.toDouble(),
                uid = info.uid,
                factorId = info.factorId))
      } else {
        pigeonHints.add(
            InternalMultiFactorInfo(
                displayName = info.displayName,
                enrollmentTimestamp = info.enrollmentTimestamp.toDouble(),
                uid = info.uid,
                factorId = info.factorId))
      }
    }
    return pigeonHints
  }

  fun multiFactorInfoToMap(hints: List<MultiFactorInfo>): List<List<Any?>> {
    val pigeonHints = ArrayList<List<Any?>>()
    for (info in multiFactorInfoToPigeon(hints)) {
      pigeonHints.add(info.toList())
    }
    return pigeonHints
  }

  fun parseActionCodeResult(actionCodeResult: ActionCodeResult): InternalActionCodeInfo {
    val operation =
        when (actionCodeResult.operation) {
          ActionCodeResult.PASSWORD_RESET -> ActionCodeInfoOperation.PASSWORD_RESET
          ActionCodeResult.VERIFY_EMAIL -> ActionCodeInfoOperation.VERIFY_EMAIL
          ActionCodeResult.RECOVER_EMAIL -> ActionCodeInfoOperation.RECOVER_EMAIL
          ActionCodeResult.SIGN_IN_WITH_EMAIL_LINK -> ActionCodeInfoOperation.EMAIL_SIGN_IN
          ActionCodeResult.VERIFY_BEFORE_CHANGE_EMAIL ->
              ActionCodeInfoOperation.VERIFY_AND_CHANGE_EMAIL
          ActionCodeResult.REVERT_SECOND_FACTOR_ADDITION ->
              ActionCodeInfoOperation.REVERT_SECOND_FACTOR_ADDITION
          else -> ActionCodeInfoOperation.UNKNOWN
        }

    val actionCodeInfo = actionCodeResult.info
    val data =
        if (actionCodeInfo != null &&
            (actionCodeResult.operation == ActionCodeResult.VERIFY_EMAIL ||
                actionCodeResult.operation == ActionCodeResult.PASSWORD_RESET)) {
          InternalActionCodeInfoData(email = actionCodeInfo.email)
        } else if (actionCodeResult.operation == ActionCodeResult.RECOVER_EMAIL ||
            actionCodeResult.operation == ActionCodeResult.VERIFY_BEFORE_CHANGE_EMAIL) {
          val actionCodeEmailInfo = actionCodeInfo as ActionCodeEmailInfo
          InternalActionCodeInfoData(
              email = actionCodeEmailInfo.email, previousEmail = actionCodeEmailInfo.previousEmail)
        } else {
          InternalActionCodeInfoData()
        }

    return InternalActionCodeInfo(operation = operation, data = data)
  }

  fun parseTokenResult(tokenResult: GetTokenResult): InternalIdTokenResult {
    return InternalIdTokenResult(
        token = tokenResult.token,
        signInProvider = tokenResult.signInProvider,
        authTimestamp = tokenResult.authTimestamp * 1000,
        expirationTimestamp = tokenResult.expirationTimestamp * 1000,
        issuedAtTimestamp = tokenResult.issuedAtTimestamp * 1000,
        claims = tokenResult.claims as Map<String?, Any?>?,
        signInSecondFactor = tokenResult.signInSecondFactor)
  }
}
